presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/BingTileFunctions.java

```diff
@@ -371,16 +394,101 @@ private static Point tileXYToLatitudeLongitude(int tileX, int tileY, int zoomLev
+        return axisToCoordinates(y, mapSize);
+    }
 
+    /**
+     * Take axis and convert it to Tile coordinates
+     *
+     * @param axis
+     * @param mapSize
+     * @return
+     */
+    private static int axisToCoordinates(double axis, int mapSize)
+    {
+        int tileAxis = (int) clip(axis * mapSize, 0, mapSize - 1);
+        return tileAxis / TILE_PIXELS;
+    }
+
+    private static boolean coordinateInsideBoundary(int coordinate, int mapSize)
+    {
+        return (coordinate >= 0 && coordinate <= (mapSize - 1) / TILE_PIXELS);
+    }
+
+    /**
+     * Returns a Bing tile at a given zoom level containing a point at a given latitude and longitude.
+     * Latitude must be within [-85.05112878, 85.05112878] range. Longitude must be within [-180, 180] range.
+     * Zoom levels from 1 to 23 are supported.
+     *
+     * @param latitude
+     * @param longitude
+     * @param zoomLevel
+     * @return
+     */
+    private static BingTile latitudeLongitudeToTile(double latitude, double longitude, int zoomLevel)
+    {
+        int mapSize = mapSize(zoomLevel);
+        int tileX = longitudeToTileX(longitude, mapSize);
+        int tileY = longitudeToTileY(latitude, mapSize);
+        return BingTile.fromCoordinates(tileX, tileY, zoomLevel);
+    }
+
+    /**
+     * Returns a collection of tiles at a given zoom level surrounding a point at a given latitude and longitude.
+     * It first locate the tile in the center.
+     * then it goes around to collect the tiles nearby
+     *
+     * @param latitude
+     * @param longitude
+     * @param zoomLevel
+     * @return
+     */
+    private static Set<BingTile> latitudeLongitudeToSurroundingTiles(double latitude, double longitude, int zoomLevel, int radius)
+    {
         int mapSize = mapSize(zoomLevel);
-        int tileX = (int) clip(x * mapSize, 0, mapSize - 1);
-        int tileY = (int) clip(y * mapSize, 0, mapSize - 1);
-        return BingTile.fromCoordinates(tileX / TILE_PIXELS, tileY / TILE_PIXELS, zoomLevel);
+        checkCondition((radius >= 0 && radius < mapSize / TILE_PIXELS), "radius should be between 0 and map size");
+
+        int tileX = longitudeToTileX(longitude, mapSize);
+        int tileY = longitudeToTileY(latitude, mapSize);
+
+        Set<BingTile> bingTiles = new HashSet<>();
+
+        IntStream.rangeClosed(-1 * radius, radius).forEach(i -> {
+            IntStream.rangeClosed(-1 * radius, radius).forEach(j -> {
+                if (coordinateInsideBoundary(tileX + i, mapSize)
+                        && coordinateInsideBoundary(tileY + j, mapSize)) {
+                    bingTiles.add(BingTile.fromCoordinates(tileX + i, tileY + j, zoomLevel));
+                }
+            });
+        });
+        return Collections.unmodifiableSet(bingTiles);
+    }
+
+    private static Set<BingTile> latitudeLongitudeToSurroundingTiles(double latitude, double longitude, int zoomLevel)
```

Comment:
Move it above `latitudeLongitudeToSurroundingTiles(double latitude, double longitude, int zoomLevel, int radius)`

Reply:
Actually it seems to be unused.

Reply:
It is been used. it default radius to 1 which is immediate neighbor tiles, called directly by bingTilesAround

Reply:
Yes, you are right. Why do you need the one that accepts a radius than?

Reply:
I added it because I feel searching nearby neighbor is a generic behavior that in future user may naturally want tiles surrounding a coordinates with a radius

Reply:
> in future user may

May or may not. It is always good to think about extensibility. But extensibility often decreases readability. Everyone who is gonna be reading this code will be answering this questions to himself `Why do we need this radius parameter?` `It seems to be unused, do we call this method dynamically?` `No, there is not MethodHandle to this method, do we call this from a generated code?` and so on.